src/Parser.php

```diff
@@ -806,104 +756,23 @@ private function isExpressionStart($token) {
-                case TokenKind::VariableName:
-                case TokenKind::DollarToken:
-                    return true;
+            if ($token->isExpressionStart()) {
+                return true;
+            }
 
-                // qualified-name
+            switch ($token->kind) {
                 case TokenKind::Name:
-                case TokenKind::BackslashToken:
+                case TokenKind::ArrayKeyword:
                     return true;
                 case TokenKind::NamespaceKeyword:
                     // TODO currently only supports qualified-names, but eventually parse namespace declarations
                     return $this->isNamespaceKeywordStartOfExpression($token);
-
-                // literal
-                case TokenKind::DecimalLiteralToken: // TODO merge dec, oct, hex, bin, float -> NumericLiteral
-                case TokenKind::OctalLiteralToken:
-                case TokenKind::HexadecimalLiteralToken:
-                case TokenKind::BinaryLiteralToken:
-                case TokenKind::FloatingLiteralToken:
-                case TokenKind::InvalidOctalLiteralToken:
-                case TokenKind::InvalidHexadecimalLiteral:
-                case TokenKind::InvalidBinaryLiteral:
-                case TokenKind::IntegerLiteralToken:
-
-                case TokenKind::StringLiteralToken:
-
-                case TokenKind::SingleQuoteToken:
-                case TokenKind::DoubleQuoteToken:
-                case TokenKind::HeredocStart:
-                case TokenKind::BacktickToken:
-
-                // array-creation-expression
-                case TokenKind::ArrayKeyword:
-                case TokenKind::OpenBracketToken:
-
-                // intrinsic-construct
-                case TokenKind::EchoKeyword:
-                case TokenKind::ListKeyword:
-                case TokenKind::UnsetKeyword:
-
-                // intrinsic-operator
-                case TokenKind::EmptyKeyword:
-                case TokenKind::EvalKeyword:
-                case TokenKind::ExitKeyword:
-                case TokenKind::DieKeyword:
-                case TokenKind::IsSetKeyword:
-                case TokenKind::PrintKeyword:
-
-                // ( expression )
-                case TokenKind::OpenParenToken:
-                case TokenKind::ArrayCastToken:
-                case TokenKind::BoolCastToken:
-                case TokenKind::DoubleCastToken:
-                case TokenKind::IntCastToken:
-                case TokenKind::ObjectCastToken:
-                case TokenKind::StringCastToken:
-                case TokenKind::UnsetCastToken:
-
-                // anonymous-function-creation-expression
-                case TokenKind::StaticKeyword:
-                case TokenKind::FunctionKeyword:
-                case TokenKind::FnKeyword:
-                    return true;
             }
-            return \in_array($token->kind, $this->reservedWordTokens, true);
+            return $token->isReserved();
```

Comment:
I assume end-users would care more about the performance of parsing than the design - function calls are fairly slow in php, and replacing in_array with a call would likely hurt performance.

Replacing with checking for set membership would be a tiny bit faster.

Also, is ArrayKeyword not reserved?

Reply:
A question this raises is whether it'd be useful to provide these helpers on Token to end-users.

What use cases did you have in mind for applications using tolerant-php-parser with isReserved()?

Reply:
I used the TokenStringMaps as reference, the same Tokens in KEYWORDS and RESERVED_WORDS are in the isKeyword() and isReserved(). 

Perhaps the Token class is not the right place to use it, at the end, in the Token class these methods are only shortcuts to TokenKind::is...(int $kind) and these may be relevant to the Parser only.

My initial goal was to rely on these methods instead of Parser's properties:
- reservedWordTokens
- keywordTokens
- nameOrKeywordOrReservedWordTokens
- nameOrReservedWordTokens
- nameOrStaticOrReservedWordTokens
- parameterTypeDeclarationTokens
- returnTypeDeclarationTokens

And shrink methods isExpressionStart and isStatementStart, which are called many times during parsing.